src/utils/url.ts

```diff
@@ -70,10 +70,23 @@ export const getPattern = (label: string): Pattern | null => {
 }
 
 export const getPath = (request: Request): string => {
-  // Optimized: indexOf() + slice() is faster than RegExp
   const url = request.url
-  const queryIndex = url.indexOf('?', 8)
-  return url.slice(url.indexOf('/', 8), queryIndex === -1 ? undefined : queryIndex)
+  const start = url.indexOf('/', 8)
+  let i = start
+  for (; i < url.length; i++) {
+    const charCode = url.charCodeAt(i)
+    if (charCode === 37) {
+      // '%'
+      // If the path contains percent encoding, use `indexOf()` to find '?' and return the result immediately.
+      // Although this is a performance disadvantage, it is acceptable since we prefer cases that do not include percent encoding.
+      const queryIndex = url.indexOf('?', i)
+      return decodeURIComponent(url.slice(start, queryIndex === -1 ? undefined : queryIndex))
```

Comment:
```suggestion
      return decodeURI(url.slice(start, queryIndex === -1 ? undefined : queryIndex))
```

Reply:
Thanks for the suggestion.

With the current code, if we `s/decodeURIComponent/decodeURI/` , the following test fails, are you saying that "id is çawa yî%3F" is the expected result?

```diff
- id is çawa yî?
+ id is çawa yî%3F
````

https://github.com/usualoma/hono/blob/29735fe9cb5038ba5a45245be93344e5c16cfaef/src/hono.test.ts#L742-L748


This can be fixed by reverting 7192497. If we do that, the following test will fail. (double decoded)

```diff
- posts of %25
+ posts of %
````

https://github.com/usualoma/hono/blob/29735fe9cb5038ba5a45245be93344e5c16cfaef/src/hono.test.ts#L768-L774

Reply:
> This can be fixed by reverting https://github.com/honojs/hono/commit/7192497e69a38e2915c84b1a4b7f80551e3e99f9.

Correct.

> If we do that, the following test will fail. (double decoded)

Ah, I see now, thanks for explaining! Well, in that case we need to `.replaceAll('%', '%25')` prior to calling `decodeURIComponent`.

Reply:
@szmarczak Thank you!

Unfortunately,

```
decodeURI('%2F') === decodeURI('%252F')
```

And cannot in principle be reverted with `replaceAll()`, since the original information is lost when the `decodeURI()` is applied.

For my part, I think that "using decodeURI() on a requested URL" is impossible.

Reply:
What about doing `.replaceAll('%25', '%2525')` before calling `decodeURI`? This means that in order to catch `%` people will need to use `%25` in the path string, e.g. `/foo%25bar` (not: `/foo%bar`), but for all other percent-encoded characters it should decode as expected.

Reply:
Ah I see, that is true. That's great! I think it will be the spec you want.

I'm not going to include this change in this PR, since this PR has the same result as Django, but I'll make a separate PR for this.

Reply:
Thank you!